drop-in/src/main/java/com/adyen/checkout/dropin/service/SessionDropInService.kt

```diff
@@ -99,11 +115,48 @@ open class SessionDropInService : DropInService(), SessionDropInServiceInterface
         }
     }
 
+    protected open fun makePaymentsCallMerchant(paymentComponentState: PaymentComponentState<*>, paymentComponentJson: JSONObject) : Boolean {
+        return false
+    }
+
+    override fun onPaymentsCallRequested(
+        paymentComponentState: PaymentComponentState<*>,
+        paymentComponentJson: JSONObject
+    ) {
+        val callWasHandled = makePaymentsCallMerchant(paymentComponentState, paymentComponentJson)
```

Comment:
I see you are doing the same in all these methods, so maybe this logic can be extracted into something like this
```
private fun cantComeUpWithAGoodName(actionComponentData, actionComponentJson, merchantCall: (actionComponentData, actionComponentJson) -> Boolean, internalCall: (actionComponentData) -> Unit) {
         val callWasHandled = merchantCall(actionComponentData, actionComponentJson)
         if (!callWasHandled) {
             if (isFlowTakenOverFlow.value) {
                 throw CheckoutException("")
             } else {
                 internalCall(actionComponentData)
             }
         } else {
             if (!isFlowTakenOverFlow.value) {
                 isFlowTakenOverFlow.value = true
             }
         }
}
```

Reply:
i wrote this function exactly, however all of these methods take different params. i can have a typed function then there are calls with 0 or 1 params instead of 2 so that means i can't use a single typed function. so unfortunately this is the way to go

Reply:
What if we use lambdas?

```kotlin
    private fun cantComeUpWithAGoodNameEither(
        makeMerchantCall: () -> Boolean,
        makeInternalCall: () -> Unit,
        merchantMethodName: String
    ) {
        val callWasHandled = makeMerchantCall()
        if (!callWasHandled) {
            if (_isFlowTakenOverFlow.value) {
                throw CheckoutException("Sessions flow was already taken over in a previous call, $merchantMethodName should be implemented")
            } else {
                makeInternalCall()
            }
        } else {
            if (!_isFlowTakenOverFlow.value) {
                _isFlowTakenOverFlow.value = true
            }
        }
    }



    override fun onPaymentsCallRequested(
        paymentComponentState: PaymentComponentState<*>,
        paymentComponentJson: JSONObject
    ) {
        cantComeUpWithAGoodNameEither(
            makeMerchantCall = {
                makePaymentsCallMerchant(paymentComponentState, paymentComponentJson)
            },
            makeInternalCall = {
                makePaymentsCallInternal(paymentComponentState)
            },
            merchantMethodName = ::makePaymentsCallMerchant.name //not sure if this will return the method name
        )
    }

```